File: pygowave_rpc/stomp_client.py

#
# PyGoWave Server - The Python Google Wave Server
# Copyright 2009 Patrick Schneider <patrick.p2k.schneider@gmail.com>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from twisted.internet.protocol import Protocol, ReconnectingClientFactory
from twisted.internet.task import LoopingCall

import stomper, anyjson

from c2s_mp import PyGoWaveClientMessageProcessor

from django.conf import settings

__all__ = ["StompClientFactory"]

class StompMessageProcessor(stomper.Engine):
	def __init__(self):
		super(StompMessageProcessor, self).__init__()
		self.username = getattr(settings, "STOMP_USER", "pygowave_server")
		self.password = getattr(settings, "STOMP_PASSWORD", "pygowave_server")
		self.pygo_mp = PyGoWaveClientMessageProcessor()
	
	def connect(self):
		"""Generate the STOMP connect command to get a session."""
		self.pygo_mp.logger.info("=> PyGoWave RPC Server starting <=")
		return stomper.connect(self.username, self.password)
	
	def connected(self, msg):
		"""Once I've connected I want to subscribe to my the message queue."""
		super(StompMessageProcessor, self).connected(msg)
		
		self.pygo_mp.logger.info("=> PyGoWave RPC Server ready <=")
		
		f = stomper.Frame()
		f.unpack(stomper.subscribe("wavelet_rpc_singlethread"))
		f.headers["exchange"] = "wavelet.topic"
		f.headers["routing_key"] = "#.#.clientop"
		f.headers["exchange_type"] = "topic"
		return f.pack()
	
	def ack(self, message):
		rkey = message["headers"]["destination"]
		message_data = anyjson.deserialize(message["body"])
		
		msg_dict = self.pygo_mp.process(rkey, message_data)
		
		out_frames = ""
		for out_rkey, messages in msg_dict.iteritems():
			out_frames += self.send(out_rkey, messages)
		
		return super(StompMessageProcessor, self).ack(message) + out_frames

	def send(self, routing_key, message_data):
		"""Convert a routing_key and data dictionary into a STOMP message."""
		f = stomper.Frame()
		f.unpack(stomper.send(routing_key, anyjson.serialize(message_data)))
		f.headers["exchange"] = "wavelet.direct"
		f.headers["content-type"] = "application/json"
		return f.pack().encode("utf-8")
	
	def purge_connections(self):
		"""Forwards to PyGoWaveClientMessageProcessor"""
		self.pygo_mp.purge_connections()

class StompClientProtocol(Protocol):
	def __init__(self):
		self.mp = StompMessageProcessor()
	
	def connectionMade(self):
		"""Register with the stomp server."""
		self.transport.write(self.mp.connect())
		self.lc = LoopingCall(self.mp.pygo_mp.purge_connections)
		self.lc.start(10 * 60) # Purge every 10 minutes
	
	def connectionLost(self, reason):
		if self.lc.running:
			self.lc.stop()
	
	def dataReceived(self, data):
		"""Data received, react to it and respond."""
		msg = stomper.unpack_frame(data)
		returned = self.mp.react(msg)
		if returned:
			self.transport.write(returned)

class StompClientFactory(ReconnectingClientFactory):
	def startedConnecting(self, connector):
		"""Started to connect."""
	
	def buildProtocol(self, addr):
		"""Transport level connected now create the communication protocol."""
		return StompClientProtocol()
	
	def clientConnectionLost(self, connector, reason):
		"""Lost connection."""
	
	#def clientConnectionFailed(self, connector, reason):
	#	"""Connection failed."""
	#	super(StompClientFactory, self).clientConnectionFailed(connector, reason)
	
	def __repr__(self):
		return "StompClientFactory"

File: pygowave_erlang_rpc/stomp_client.erl

%%
%% PyGoWave Server - The Python Google Wave Server
%% Copyright 2009 Patrick Schneider <patrick.p2k.schneider@gmail.com>
%%
%% Licensed under the Apache License, Version 2.0 (the "License");
%% you may not use this file except in compliance with the License.
%% You may obtain a copy of the License at
%%
%%     http://www.apache.org/licenses/LICENSE-2.0
%%
%% Unless required by applicable law or agreed to in writing, software
%% distributed under the License is distributed on an "AS IS" BASIS,
%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
%% See the License for the specific language governing permissions and
%% limitations under the License.
%%

-module(stomp_client).
-export([
	start/0,
	msg_sender/1,
	msg_receiver/3,
	run_purge/1
	]).

%% Defines
-include("settings.hrl").

%% Logging
log(Module, Line, Level, Format, Arguments) ->
	if
		(Module == mysql) or (Module == mysql_conn) or (Module == mysql_auth) ->
			ok;
		true ->
			LogOk = case ?LOG_LEVEL of
				debug ->
					true;
				normal ->
					case Level of
						debug -> false;
						_ -> true
					end;
				error ->
					case Level of
						debug -> false;
						normal -> false;
						_ -> true
					end
			end,
			case LogOk of
				true -> io:format("~w:~b: " ++ Format ++ "~n", [Module, Line] ++ Arguments);
				_ -> ok
			end
	end.

-define(Log(Level, Format, Arguments),
	log(?MODULE, ?LINE, Level, Format, Arguments)).

log(Module, Line, Level, FormatFun) ->
	{Format, Arguments} = FormatFun(),
	log(Module, Line, Level, Format, Arguments).

start() ->
	%% Connect to database
	case mysql:start_link(pygo_sql, "localhost", undefined, "pygowave", "pygowave", "pygowave", fun log/4, utf8) of
		{ok, _} -> ok;
		{error, {already_started, _}} -> ok
	end,
	%% Connect to message queue
	Conn = stomp:connect("localhost", 61613, "pygowave_server", "pygowave_server"),
	%% Subscribe to queue
	stomp:subscribe("wavelet_rpc_singlethread", Conn, [{"exchange", "wavelet.topic"}, {"routing_key", "#.#.clientop"}, {"exchange_type", "topic"}]),
	%% Spawn Message Processor
	MP = spawn(c2s_mp, start, [pygo_sql]),
	%% Spawn Message Sender
	S = spawn(?MODULE, msg_sender, [Conn]),
	%% Initial purge & setup purge timer (every 10 minutes)
	MP ! purge,
	{ok, _PurgeTimer} = timer:apply_interval(600000, ?MODULE, run_purge, [MP]),
	%% Enter message loop
	stomp:on_message(fun (Frame) -> msg_receiver(MP, S, Frame) end, Conn).

run_purge(MP) ->
	MP ! purge.

%% Receiver
msg_receiver(MP, S, Frame) ->
	case read_frame(Frame) of
		{ok, RKey, Data} ->
			?Log(debug, "Received from ~s: ~s", [RKey, rfc4627:encode(Data)]),
			MP ! {process, S, RKey, Data}, ok;
		{error, _} -> error
	end.

%% Sender
msg_sender(Conn) ->
	receive
		Msgs -> msg_sender(Conn, Msgs)
	end,
	msg_sender(Conn).

msg_sender(_, []) ->
	ok;
msg_sender(Conn, [{RKey, Msgs}|T]) ->
	EncMsgs = rfc4627:encode(Msgs),
	?Log(debug, "Sending to   ~s: ~s", [RKey, EncMsgs]),
	stomp:send(
		Conn,
		RKey,
		[{"exchange","wavelet.direct"},{"content-type","application/json"},{"content-encoding","utf-8"}],
		EncMsgs
	),
	msg_sender(Conn, T).

%% Read and process a STOMP frame
read_frame([{type, Type}, {headers, Headers}, {body, Body}]) ->
	read_frame(Type, Headers, Body);
read_frame(_) ->
	?Log(error, "Got bogus data from STOMP library!", []),
	{error, bogus_data}.

read_frame("ESSAGE", Headers, Body) ->
	read_frame("MESSAGE", Headers, Body);

read_frame("MESSAGE", Headers, Body) ->
	case conv_headers(Headers) of
		[{destination, Dest},{message_id, _},{content_type, "application/json"},{content_encoding, "utf-8"}] ->
			parse_message(Dest, Body);
		[{destination, _},{message_id, _},{content_type, CType},{content_encoding, CEnc}] ->
			?Log(error, "Unknown content-type \"~s\" and/or encoding \"~s\", skipping.", [CType, CEnc]),
			{error, wrong_type_or_encoding}
	end;
read_frame(Type, _, _) ->
	?Log(error, "Unknown STOMP frame type \"~s\", skipping.", [Type]),
	{error, unknown_stomp_frame}.

%% Parse the headers
conv_headers(Headers) ->
	conv_headers(Headers, [{destination, ""},{message_id, ""},{content_type, ""},{content_encoding, ""}]).

conv_headers([], CvdHeaders) ->
	CvdHeaders;
conv_headers([H|T], [{destination, Dest},{message_id, MsgId},{content_type, CType},{content_encoding, CEnc}]) ->
	case H of
		{"destination", Val} -> conv_headers(T, [{destination, Val},{message_id, MsgId},{content_type, CType},{content_encoding, CEnc}]);
		{"message-id", Val} -> conv_headers(T, [{destination, Dest},{message_id, Val},{content_type, CType},{content_encoding, CEnc}]);
		{"content-type", Val} -> conv_headers(T, [{destination, Dest},{message_id, MsgId},{content_type, Val},{content_encoding, CEnc}]);
		{"content-encoding", Val} -> conv_headers(T, [{destination, Dest},{message_id, MsgId},{content_type, CType},{content_encoding, Val}]);
		_ -> conv_headers(T, [{destination, Dest},{message_id, MsgId},{content_type, CType},{content_encoding, CEnc}])
	end.

%% Parse a message
parse_message(Dest, Body) ->
	case rfc4627:decode(Body) of
		{ok, Data, []} -> {ok, Dest, Data};
		{ok, _, _} -> {error, json_has_remainder};
		{error, Reason} -> {error, {json_error, Reason}}
	end.
